py/stabilizer/stream.py

```diff
@@ -1,92 +1,159 @@
+        # convert DAC offset binary to two's complement
+        data[2:] ^= np.int16(0x8000)
+        return data
 
+    def to_si(self):
+        """Convert the raw data to SI units"""
+        data = self.to_mu() * DAC_VOLTS_PER_LSB
+        return {
+            "adc": data[:2],
+            "dac": data[2:],
+        }
 
-    def get_rx_bytes(self):
-        """ Get the number of bytes read from the stream. """
-        return self.total_bytes
 
+class StabilizerStream(asyncio.DatagramProtocol):
+    """Stabilizer streaming receiver protocol"""
+    # The magic header half-word at the start of each packet.
+    magic = 0x057B
+    header_fmt = struct.Struct("<HBBI")
+    header = namedtuple("Header", "magic format_id batch_size sequence")
+    parsers = {
+        AdcDac.format_id: AdcDac,
+    }
 
-    def read_frame(self):
-        """ Read a single frame from the stream.
+    @classmethod
+    async def open(cls, local_addr, maxsize=1):
+        """Open a UDP socket and start receiving frames"""
+        loop = asyncio.get_running_loop()
+        _transport, protocol = await loop.create_datagram_endpoint(
+            lambda: cls(maxsize), local_addr=local_addr)
+        return protocol
 
-        Returns:
-            Yields the (seqnum, data) of the batches available in the frame.
-        """
-        buf = self.socket.recv(4096)
-        self.total_bytes += len(buf)
+    def __init__(self, maxsize):
+        self.queue = asyncio.Queue(maxsize)
 
-        # Attempt to parse a block from the buffer.
-        if len(buf) < struct.calcsize(self.HEADER_FORMAT):
-            return
+    def connection_made(self, _transport):
+        logger.info("Connection made (listening)")
 
-        # Parse out the packet header
-        magic, format_id, batch_size, sequence_number = struct.unpack_from(self.HEADER_FORMAT, buf)
-        buf = buf[struct.calcsize(self.HEADER_FORMAT):]
+    def connection_lost(self, _exc):
+        logger.info("Connection lost")
 
-        if magic != self.MAGIC_HEADER:
-            logging.warning('Encountered bad magic header: %s', hex(magic))
+    def datagram_received(self, data, _addr):
+        header = self.header._make(self.header_fmt.unpack_from(data))
+        if header.magic != self.magic:
+            logger.warning("Bad frame magic: %#04x, ignoring", header.magic)
             return
-
-        frame_format = self.FORMAT[format_id](batch_size)
-
-        batch_count = int(len(buf) / struct.calcsize(frame_format))
-
-        for offset in range(batch_count):
-            data = struct.unpack_from(frame_format, buf)
-            buf = buf[struct.calcsize(frame_format):]
-            yield (sequence_number + offset, data)
+        try:
+            parser = self.parsers[header.format_id]
+        except KeyError:
+            logger.warning("No parser for format %s, ignoring", header.format_id)
+            return
+        frame = parser(header, data[self.header_fmt.size:])
```

Comment:
My only qualm with this approach is that this implies that the user gets a variable type back from the `StabilizerStream.queue`. Essentially, there's no guarantee on what the type is, and the type changes based on the format_id of the received frame.

In Rust, we'd solve this by returning some `impl StreamFrame`. In our case, I guess we could use an ABC to guarantee that the returned object implements some common interface?

We need to have some known API so something like the stream visualizer can know what to do for any arbitrary stream frame received.

Reply:
I'm not sure there is a common meaningful `impl StreamFrame` trait. Apart from the header data, what's guaranteed to be common among stream frame formats?

The user can use any of the following approaches:

* adjust `parsers` so they are sure to only receive what they want
* check the `.header.format_id` attribute of the object they get form the queue
* use `is AdcDac` to check for class
* use duck typing to check for behavioral equivalence